File: online judge.kaner murillo/python/11235.py
#https://onlinejudge.org/external/112/11235.pdf
#11235

from sys import stdin

def find(A, tree, l, r, position, val):

    ans = None

    if l == r:

        ans = 0
        
        if A[r] == val:

            ans = 1

    else:

        mid = l + ((r-l)>>1)

        ans = find(A, tree, l, mid, position+1, val) + find(A, tree, mid+1, r, position+2*(mid-l+1), val)
    
    return ans

def build(A, tree, auxTree, l, r, position):


    if l == r:

        tree[position] = (A[r], 1)

    else:

        mid = l + ((r-l)>>1)

        build(A, tree, auxTree, l, mid, position+1)
        build(A, tree, auxTree, mid+1, r, position+2*(mid-l+1))

        if tree[position+1][0] == tree[position+2*(mid-l+1)][0]:

            tree[position] = (tree[position+1][0], tree[position+1][1]+tree[position+2*(mid-l+1)][1])

        else:
            
            if tree[position+1][1] > tree[position+2*(mid-l+1)][1]:

                aux = find(A, auxTree, mid+1, r, position+2*(mid-l+1), tree[position+1][0])
                tree[position] = (tree[position+1][0], tree[position+1][1] + aux)

            elif tree[position+1][1] < tree[position+2*(mid-l+1)][1]:
                
                aux = find(A, auxTree, l, mid, position+1, tree[position+2*(mid-l+1)][0])
                tree[position] = (tree[position+2*(mid-l+1)][0], tree[position+2*(mid-l+1)][1] + aux)

            else:

                aux1 = find(A, auxTree, mid+1, r, position+2*(mid-l+1), tree[position+1][0])
                aux2 = find(A, auxTree, l, mid, position+1, tree[position+2*(mid-l+1)][0])

                if tree[position+2*(mid-l+1)][1] + aux2 > tree[position+1][1] + aux1:

                    tree[position] = (tree[position+2*(mid-l+1)][0], tree[position+2*(mid-l+1)][1] + aux2)                

                else:
                    
                    tree[position] = (tree[position+1][0], tree[position+1][1] + aux1)

def querie(tree, auxTree, L, R, l, r, position):

    ans = None

    if l>r:

        ans = (float("-inf"), float("-inf"))
    
    elif L == l and R == r:

        ans = tree[position]
        print(ans)

    else:

        mid = L+((R-L)>>1)

        ans1 = querie(tree, auxTree, L, mid, l, min(mid, r), position+1)
        ans2 = querie(tree, auxTree, mid+1, R, max(l, mid+1), r, position+2*(mid-L+1))

        print(ans1, ans2)
        
        if ans1[0] == ans2[0]:
            ans = (ans1[0], ans1[1]+ans2[1])

        elif ans1[1] > ans2[1]:
            ans = ans1 
        
        else:

            ans = ans2

    return ans

def main():

    case = [int(x) for x in stdin.readline().split()]

    while len(case) > 1:

        A = [int(x) for x in stdin.readline().split()]
        tree = [0 for _ in range(len(A)*2)]
        auxTree = [0 for _ in range(len(A)*2)]
        build(A, tree, auxTree, 0, case[0]-1, 0)
        
        print(tree)
        print(auxTree)

        for i in range(case[1]):

            querie1 = [int(x) for x in stdin.readline().split()]

            print(querie(tree, auxTree, 0, case[0]-1, querie1[0]-1, querie1[1]-1, 0)[1])
            
        case = [int(x) for x in stdin.readline().split()]

main()

File: online judge.kaner murillo/c++/11235.cpp
//https://onlinejudge.org/external/112/11235.pdf
//11235

#include<cstdio>
#include<iostream>
#include<vector>

using namespace std;

int tree[100001*2], frequent[100001], A[100001], start[100001];

void build(int l, int r, int pos){

    if(l == r){
        tree[pos] = frequent[l];
    }

    else{
        int mid = l + ((r-l)>>1);



        build(l, mid, pos+1);
        build(mid+1, r, pos+2*(mid-l+1));
            
        tree[pos] = max(tree[pos+1], tree[pos+2*(mid-l+1)]);
    }   
}

int querie(int L, int R, int l, int r, int pos){

    if(l>r){
        return -1;
    }

    else if(l == L && r == R){
        return tree[pos];
    }

    else{

        int mid = L + ((R-L)>>1);
        return max(querie(L, mid, l, min(mid, r), pos+1), querie(mid+1, R, max(mid+1, l), r, pos+2*(mid-L+1)));
    }
}

int main(){

    int size, queries, l, r, l2, r2, i, j, index, val, preVal, size1, size2, counter;

    while(scanf("%i", &size) && size){
        scanf("%i", &queries);
        counter = 1;
        index = 0;
        for(i = 0; i<size; i++){
            scanf("%i", &val);
            A[i] = val;
            if(i){
                if(val == preVal){
                    counter++;
                }    

                else{
                    for(j = index; j<index+counter; j++){
                        frequent[j] = counter;
                        start[j] = index;
                    } 
                    index = index+counter;
                    counter = 1;
                }
            }

            preVal = val;
        }

        for(j = index; j<index+counter; j++){
            frequent[j] = counter;
            start[j] = index;
        } 

        build(0, size-1, 0);

        for(i = 0; i<queries; i++){
            scanf("%i %i", &l, &r);
            l--, r--;

            if(A[l] == A[r]){
                printf("%i\n", r-l+1);
            }
            else{

                size1 = frequent[l] - (l - start[l]);
                size2 = r - start[r] + 1;

                if(size1 + size2 == r-l+1){
                    printf("%i\n", max(size1, size2));
                }

                else{
                    printf("%i\n", max(max(size1, size2), querie(0, size-1, l+size1, r-size2, 0)));
                }

            }
        }
    }

    return 0;
}